edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/storage/MetricsStorage.cs

```diff
@@ -27,23 +31,54 @@ public Task StoreMetricsAsync(IEnumerable<Metric> metrics)
 
         public async Task<IEnumerable<Metric>> GetAllMetricsAsync()
         {
-            return (await this.dataStore.GetBatch(0, 1000))
-                    .SelectMany(((long offset, IEnumerable<Metric> metrics) storeResult) =>
+            // Note: async/await is needed b/c ToListAsync returns ValueTask instead of Task.
+            return await this.GetAllMetricsInternal().ToListAsync();
+        }
+
+        async IAsyncEnumerable<Metric> GetAllMetricsInternal([EnumeratorCancellation] CancellationToken cancellationToken = default)
+        {
+            HashSet<long> maxAgeEntriesToRemove = new HashSet<long>();
+
+            foreach ((long offset, IEnumerable<Metric> metrics) in await this.dataStore.GetBatch(0, 1000, cancellationToken))
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                // This makes the assumption that metrics are stored in batches with matching timestamps.
+                // This is the case b/c the metrics are stored immediately after being scraped.
+                if (metrics.First().TimeGeneratedUtc < DateTime.UtcNow - this.maxMetricAge)
+                {
+                    // TODO: log removed
+                    maxAgeEntriesToRemove.Add(offset);
+                }
+                else
+                {
+                    this.offsetsReturned.Add(offset);
+                    foreach (var metric in metrics)
                     {
-                        this.entriesToRemove.Add(storeResult.offset);
-                        return storeResult.metrics;
-                    });
+                        yield return metric;
+                    }
+                }
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // remove metrics past ttl
+            await this.RemoveOffsets(maxAgeEntriesToRemove);
+        }
+
+        public Task RemoveAllReturnedMetricsAsync()
```

Comment:
Remove the `Async` suffix?

Reply:
Its still a task, I thought we named it async if it returns a task. Thats all the caller cares about anyway.

Reply:
You're right. Not sure what I was thinking...